examples/Spirit/spirit_utils.cc

```diff
@@ -86,6 +86,113 @@ void nominalSpiritStand( MultibodyPlant<T>& plant, Eigen::VectorXd& xState, doub
   }
 }
 
+void ikSpiritStand(drake::multibody::MultibodyPlant<double>& plant,
+                   Eigen::VectorXd& xState,
+                   Eigen::Matrix<bool,1,4> contactSequence,
+                   double com_height,
+                   double leg_height,
+                   double roll,
+                   double pitch,
+                   double eps){
+  drake::multibody::InverseKinematics ik(plant);
+  const auto& world_frame = plant.world_frame();
+  const auto& body_frame  = plant.GetFrameByName("body");
+
+  // Position of the com
+  Vector3d pos = {0, 0, com_height};
+  // Orientation of body
+  drake::math::RotationMatrix<double> orientation;
+  orientation = drake::math::RotationMatrix<double>::MakeXRotation(-roll) *
+                drake::math::RotationMatrix<double>::MakeYRotation(-pitch) ;
+
+  const double n_x = plant.num_positions() + plant.num_velocities();
+  const double n_q = plant.num_positions();
+
+
+  // Constrain body location and pose
+  ik.AddPositionConstraint(body_frame, Vector3d(0, 0, 0), world_frame,
+                           pos,
+                           pos);
+  ik.AddOrientationConstraint(body_frame, orientation,
+                              world_frame, drake::math::RotationMatrix<double>(), 0);
+
+  // Constrain toes
+  for(int toe = 0; toe < 4; toe ++){
+    constrainToe(plant, ik, toe, contactSequence[toe], orientation, com_height, leg_height, eps);
+  }
+
+  // Add initial guess and solve
+  Eigen::VectorXd initial_guess;
+  dairlib::nominalSpiritStand(plant, initial_guess, com_height);
+  ik.get_mutable_prog()->SetInitialGuess(ik.q(), initial_guess.head(n_q));
+  const auto result = Solve(ik.prog());
+  const auto q_sol = result.GetSolution(ik.q());
+  std::cout << (result.is_success() ? "IK Success" : "IK Fail") << std::endl;
+
+  // Return
+  VectorXd x_const = VectorXd::Zero(n_x);
+  x_const.head(n_q) = q_sol;
+  xState = x_const;
+}
+
+void constrainToe(drake::multibody::MultibodyPlant<double> & plant,
+                  drake::multibody::InverseKinematics &ik,
+                  int toe, bool inContact, drake::math::RotationMatrix<double> orientation,
+                  double com_height, double leg_height, double eps){
+  // Get frames
+  const auto& toe_frame = dairlib::getSpiritToeFrame(plant, toe);
+  const auto& world_frame = plant.world_frame();
+  const auto& body_frame  = plant.GetFrameByName("body");
+  Vector3d eps_vec = {eps, eps, eps};
+
+  // Chose distance between toe and com
+  const double hip_width = 0.07;
+  const double body_length = 0.2263;
+  Vector3d toe_offset;
+  switch (toe) {
+    case 0:
```

Comment:
Not sure I agree with this way of doing this over something a little more D.R.Y. but it's not worth messing with.